### ISSUE TITLE
Question on Fusepy memory example

### ISSUE THREAD

**Issue by blackening** [billziss-gh/fusepy#2]
I tried the memory example together with winfsp 1.017 on win10. I cannot get it to work.

For example, i cannot copy anything in (not enough space?) When i create a new text document, it can neither be renamed nor accessed.

I'm not sure how Windows determines whether there is enough space for a file. I noticed it runs a statfs call, but adding a few zeroes to the default `dict(f_bsize=512, f_blocks=4096, f_bavail=2048)` doesn't change anything.

However, i have determined that when Windows is attempting to access a file, it calls getattr with the filename in all caps. Can i check if this is something the person implementing the FUSE is supposed to deal with?

Thanks in advance.

**Comment by billziss-gh** [comment]
@blackening a couple of questions:

- Are you using this under Cygwin or Windows?

- Have you installed the package using `python setup.py install` or are you trying to run things from the source directory?

- Are you  mounting on a drive (`X:`) or directory (`\directory\path`)?

> However, i have determined that when Windows is attempting to access a file, it calls getattr with the filename in all caps. Can i check if this is something the person implementing the FUSE is supposed to deal with?

Windows does not handle well case-sensitive file systems mounted on directories. It tends to send them file names with mixed case (e.g. sometimes `Filename` and sometimes `FILENAME`). If you are mounting on a directory, then try mounting the file system on a drive:

    > python memory.py x:

Please also note that the fusepy `memory` file system is very simple and "supports only one level of files".

**Comment by blackening** [comment]
I don't have cygwin installed, pure win10.

Ok, i didn't set it up and tried to run it from the source directory. 

Its mounted on a directory, i used it without subfolders. Thanks for confirming the windows quirk of occasionally using mixed case and upper case. 


------------



I just tried it out. So if install it, it acts pretty much the same for copying stuff in, but i accidentally managed to move an item in, effectively deleting it X.X

Other than that, mounting it as a drive produces a 0 bytes out of 0 bytes drive.


------------
I also realized another asymmetry. You made a few changes in memory.py which used `uid, gid, pid = fuse_get_context()`
Apparently this doesn't apply to the base folder during `__init__`, is this ok? Also fuse_get_context() doesn't really work (returns 0,0,-1, from memory) unless its called after `__init__`, probably in some FUSE related callback(?).

**Comment by billziss-gh** [comment]
> I just tried it out. So if install it, it acts pretty much the same for copying stuff in, but i accidentally managed to move an item in, effectively deleting it X.X

I do not quite understand. Are you saying that you were copying a file and it got moved/deleted instead?

> Other than that, mounting it as a drive produces a 0 bytes out of 0 bytes drive.

I modified `statfs` as follows and the drive size is reported correctly. I need to make sure that WinFsp-FUSE is not doing the wrong thing here with `f_frsize` vs `f_bsize`:

```python
    def statfs(self, path):
        return dict(f_frsize=512, f_blocks=4096, f_bfree=2048)
```

> You made a few changes in memory.py which used uid, gid, pid = fuse_get_context().

During creation of files/directories the uid, gid sent from `fuse_get_context` is valid. I simply used them to make sure that I create files/directories with the correct uid/gid.

**Comment by blackening** [comment]
> I do not quite understand. Are you saying that you were copying a file and it got moved/deleted instead?


No, i dragged it in, what happens is that apparently doesn't require any space in the destination directory. Thus it got moved into a directory with 0kb of space. Then it cannot be accessed because memory.py is case sensitive and windows addresses drives having 0kb with caps.

Funnily enough, after finding out that my understanding of the parametization of statfs (and the default) was incorrect, once the drive has enough capacity, windows starts querying the drive without uppercasing everything.

Basically f_bsize is apparently not used for calculating free space? 

Anyway thanks very much, i should be able to handle everything else from here (at least i hope so).

**Comment by billziss-gh** [comment]
> Funnily enough, after finding out that my understanding of the parametization of statfs (and the default) was incorrect, once the drive has enough capacity, windows starts querying the drive without uppercasing everything.

It will unfortunately still do so on some occasions. The only way to have a really working file system mounted on a directory is to make the file system case insensitive.

> Basically f_bsize is apparently not used for calculating free space?

I may actually consider this a bug in WinFsp-FUSE. It should probably honor the `f_bsize` rather than the `f_frsize`. Although it is a rather weird file system that has `b_frsize != 0` and `b_bsize == 0` (or vice-versa.

I recommend setting both `b_bsize` and `b_frsize`. If you do not have good values in your file system, a value of `4096` (page size) is actually the best.

> Anyway thanks very much, i should be able to handle everything else from here (at least i hope so).

Please let me know if you have additional questions.

**Comment by blackening** [comment]
Sorry, i have another question. Does winfsp feed back the file handles?
For example on calls like truncate, release, read, write, etc the last parameter seems to always be provided as 0 regardless of the return value of open/create.

**Comment by billziss-gh** [comment]
WinFsp-FUSE definitely passes back the file handles from `open` during `read`, `write`. For `truncate` it depends and you may or may not get the handles.

I do not believe that the Windows fork of FUSEPY changes anything in how file handles are passed.

**Comment by blackening** [comment]
I just tested the code on Ubuntu, the handles are passed perfectly well, but not on windows. Its also pretty obvious on the logs with memory.py - return values of open/create are incremental while the calls to read/write/flush/truncate/release all have 0L as their last parameter.

Honestly, i don't see any changes in the commit log that affects the file handles, so i'm not exactly sure where it originates from.

**Comment by billziss-gh** [comment]
@blackening I will try to test it myself if I get some time today. Are you seeing this on native Windows python or Cygwin?

**EDIT**: here is WinFsp passing the `fuse_file_info` during a `read` operation:

https://github.com/billziss-gh/winfsp/blob/master/src/dll/fuse/fuse_intf.c#L1078

**Comment by blackening** [comment]
Windows native python (Python 2.7.10 64bit). I don't have cygwin installed.

Unfortunately i'm not skilled at low level C interop, if not i would have dug through the stuff already.

Thanks very much for the support.

**Comment by billziss-gh** [comment]
I can confirm that this is happening for me as well on Win10 with native Python 2.7.13. I do not know why yet.

**Comment by billziss-gh** [comment]
I have found the problem:

https://github.com/billziss-gh/fusepy/blob/master/fuse.py#L347-L357

```Python
class fuse_file_info(Structure):
    _fields_ = [
        ('flags', c_int),
        ('fh_old', c_ulong),
        ('writepage', c_int),
        ('direct_io', c_uint, 1),
        ('keep_cache', c_uint, 1),
        ('flush', c_uint, 1),
        ('padding', c_uint, 29),
        ('fh', c_uint64),
        ('lock_owner', c_uint64)]
```

If I change the `fh_old` field to be `c_int` it seems to work. However I am a bit perplexed as this should work as is on native Windows (where `unsigned long` is only 32-bit even on 64-bit Windows).

**Comment by billziss-gh** [comment]
Commit eab1b2aca692e84c73fa9e418fd058b6bf8a15d3 (currently on the [windows-fileinfo-fix](https://github.com/billziss-gh/fusepy/tree/windows-fileinfo-fix) branch) fixes this issue.

Please confirm and I will merge into the [windows](https://github.com/billziss-gh/fusepy/tree/windows) branch.

**Comment by blackening** [comment]
Yep, its working now. Thanks very much for the great support 👍

**Comment by billziss-gh** [comment]
I have now merged this into the [windows](https://github.com/billziss-gh/fusepy/tree/windows) branch. I cannot currently easily merge it into [master](https://github.com/billziss-gh/fusepy/tree/master), because there is a pending pull request from it onto the upstream project.
